File: python/src/submit_peer_to_peer_transaction_example.py
import time

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from libra import stdlib, utils, libra_types, testnet, AuthKey
from libra.testnet import CHAIN_ID

CURRENCY = "Coin1"

"""
submit_peer_to_peer_transaction_example demonstrates currencies transfer between 2 accounts on the Libra blockchain
"""


def main():
    # connect to testnet
    client = testnet.create_client()

    # generate private key for sender account
    sender_private_key = Ed25519PrivateKey.generate()
    # generate auth key for sender account
    sender_auth_key = AuthKey.from_public_key(sender_private_key.public_key())
    print(f"~ Generated sender address: {utils.account_address_hex(sender_auth_key.account_address())}")
    # create sender account
    faucet = testnet.Faucet(client)
    testnet.Faucet.mint(faucet, sender_auth_key.hex(), 1340000000, "Coin1")

    # get sender account
    sender_account = client.get_account(sender_auth_key.account_address())

    # generate private key for receiver account
    receiver_private_key = Ed25519PrivateKey.generate()
    # generate auth key for receiver account
    receiver_auth_key = AuthKey.from_public_key(receiver_private_key.public_key())
    print(f"~ Generated receiver address: {utils.account_address_hex(receiver_auth_key.account_address())}")

    # create receiver account
    faucet = testnet.Faucet(client)
    testnet.Faucet.mint(faucet, receiver_auth_key.hex(), 10000000, CURRENCY)

    # create script
    script = stdlib.encode_peer_to_peer_with_metadata_script(
        currency=utils.currency_code(CURRENCY),
        payee=receiver_auth_key.account_address(),
        amount=130000000,
        metadata=b'',  # no requirement for metadata and metadata signature
        metadata_signature=b'',
    )
    # create transaction
    raw_transaction = libra_types.RawTransaction(
        sender=sender_auth_key.account_address(),
        sequence_number=sender_account.sequence_number,
        payload=libra_types.TransactionPayload__Script(script),
        max_gas_amount=1_000_000,
        gas_unit_price=0,
        gas_currency_code=CURRENCY,
        expiration_timestamp_secs=int(time.time()) + 30,
        chain_id=CHAIN_ID,
    )
    # sign transaction
    signature = sender_private_key.sign(utils.raw_transaction_signing_msg(raw_transaction))
    public_key_bytes = utils.public_key_bytes(sender_private_key.public_key())

    signed_txn = utils.create_signed_transaction(raw_transaction, public_key_bytes, signature)
    # submit transaction
    client.submit(signed_txn)
    # wait for transaction
    client.wait_for_transaction(signed_txn)


if __name__ == "__main__":
    main()

File: typescript/src/submitPeerToPeerTransactionExample.ts
import {Signer} from "@libra/client/dist/utils/signer";
import AccountKeys from "@libra/client/dist/account/accountKeys";
import {LibraUtils, TestnetClient} from "@libra/client";
import {bytesToHexString} from "@libra/client/dist/utils/bytes";
import {Stdlib} from "@libra/client/dist/lcs/libraStdlib";
import {ChainId, RawTransaction, TransactionPayloadVariantScript} from "@libra/client/dist/lcs/libraTypes";
import {LcsSerializer} from "@libra/client/dist/lcs/lcs/lcsSerializer";
import {delay} from "./eventsSubscriber";

const CURRENCY = "Coin1";

/**
 * submitPeerToPeerTransactionExample demonstrates currencies transfer between 2 accounts on the Diem blockchain
 */
async function main() {
  //generate sender keys
  const senderKeyPair = Signer.generateKeyPair();
  const senderAccountKeys = new AccountKeys(senderKeyPair);

  //connect to client
  const client = new TestnetClient();

  //create sender account
  const senderAuthKeyHex = bytesToHexString(senderAccountKeys.authKey);
  await client.mint(senderAuthKeyHex, BigInt(1340000000), CURRENCY)
  const senderAccountAddress = senderAccountKeys.accountAddress;
  const senderAccountAddressHex = bytesToHexString(senderAccountAddress);
  console.log("~ Sender address: %s", senderAccountAddressHex)

  //get sender account
  let senderAccount = await client.getAccount(senderAccountAddress)

  let retries = 0;
  const maxRetries = 25;

  while (!senderAccount && retries < maxRetries) {
    console.log("~ no account was found for address %s retry #%i", senderAccountAddressHex, retries + 1)
    retries++;

    await delay(2);

    senderAccount = await client.getAccount(senderAccountAddress)
  }

  //generate receiver keys
  const receiverKeyPair = Signer.generateKeyPair();
  const receiverAccountKeys = new AccountKeys(receiverKeyPair);

  //create receiver account
  const receiverAuthKey = receiverAccountKeys.authKey;
  const receiverAuthKetHex = bytesToHexString(receiverAuthKey);
  await client.mint(receiverAuthKetHex, BigInt(10000000), CURRENCY)
  const receiverAccountAddressHex = bytesToHexString(receiverAccountKeys.accountAddress);
  console.log("~ Receiver address: %s", receiverAccountAddressHex)

  //create script
  console.log("~ create script")
  const [metadata, metadataSignature] = LibraUtils.createGeneralMetadata(null, null, null);

  const currencyTag = LibraUtils.makeCurrencyTypeTag(CURRENCY);

  const script = Stdlib.encodePeerToPeerWithMetadataScript(
    currencyTag,
    LibraUtils.makeAccountAddress(receiverAccountAddressHex),
    BigInt(130000000),
    metadata,
    metadataSignature
  );

  //create transaction
  const expirationTime = BigInt(Math.trunc(new Date().getTime() / 1000 + 60 * 100));

  const senderSequenceNumber = senderAccount.sequence_number;
  console.log("~ senderSequenceNumber: %i", senderSequenceNumber);

  const transaction = new RawTransaction(
    LibraUtils.makeAccountAddress(senderAccountAddressHex),
    senderSequenceNumber,
    new TransactionPayloadVariantScript(script),
    BigInt(1000000),
    BigInt(0),
    CURRENCY,
    expirationTime,
    new ChainId(2)
  );

  //sign transaction
  const signedTransaction = LibraUtils.signTransaction(transaction, senderAccountKeys);

  const signedTxSerializer = new LcsSerializer();
  signedTransaction.serialize(signedTxSerializer);

  const signedBytes = signedTxSerializer.getBytes();
  const hexTransaction = bytesToHexString(signedBytes);

  //submit transaction
  console.log("~ submit transaction")
  try {
    await client.submitRawSignedTransaction(hexTransaction);
  } catch (e) {
    console.error(e)
  }
  //wait for transaction
  console.log("~ wait for transaction")
  const transactionHash = LibraUtils.hashTransaction(signedBytes);

  try {
    await client.waitForTransaction(senderAccountAddress, senderSequenceNumber, false, transactionHash, expirationTime)
  } catch (e) {
    console.error(e)
  }
}

main()
